pkg/cidr/cidr.go

```diff
@@ -135,3 +135,55 @@ func MustParseCIDR(str string) *CIDR {
 	}
 	return c
 }
+
+// Version returns the IP version for an IP, or 0 if the IP is not valid.
+func IPVersion(i *net.IP) int {
+	if i.To4() != nil {
+		return 4
+	} else if len(*i) == net.IPv6len {
+		return 6
+	}
+	return 0
+}
+
+// matchCIDRs matchs an IP address against a list of cidrs.
+// If the list is empty, it always matches.
+func matchCIDRs(ip net.IP, cidrs []*net.IPNet) bool {
+	if len(cidrs) == 0 {
+		return true
+	}
+	for _, cidr := range cidrs {
+		if cidr.Contains(ip) {
+			return true
+		}
+	}
+	return false
+}
+
+func DetectCIDRByCIDR(cidr *CIDR, version int) (*net.Interface, net.IP, *net.IPNet, error) {
```

Comment:
The users of this function only check the specific IP and the error, so overall this function seems to be attempting the same process as `firstGlobalAddr()` from `pkg/node`, [see here](https://github.com/cilium/cilium/blob/265d6775c2e0eb4e9d1dfef3ebeaa19a9b976e5f/pkg/node/address_linux.go#L34). Is there anything about the behaviour of that function which is incorrect compared to what this function is trying to achieve? I'd like to fix up the one implementation and reuse it rather than reimplementing similar functionality in multiple places if possible.

Reply:
Trevor>> I checked firstGlobalAddr(), but it is for finding a preferredIP in a certain interface but not for matching CIDR. I think the matching/detecing CIDR functions are still natural and should not causing any bothering here.
Anyway, I would check to re-use the existing functions again if possible.